### PR TITLE
Annotating Null Island also annotates its antipode

### PR DESCRIPTION
**Pull request by brunoabinader**
If I add an annotation at the coordinates 0°, 0°, an annotation also appears at the coordinates 0°, 180°, but only at z1.

antipodes

### CODE REVIEW

**Comment by 1ec5** [comment]
Please avoid commandeering an issue opened by someone else by converting it into a PR. The resulting page is confusing for anyone who hasn’t been following the conversation, moreso if for whatever reason the PR gets rejected.

**Comment by jfirebaugh** [comment]
I'm not sure this is the right fix. Per #5419 and #1673, we actually _want_ the annotation on the border of a tile to appear in all tiles that share that border. So isn't the issue rather that the subsequent projection math places the feature at the wrong tile coordinate for some of the tiles?

**Comment by brunoabinader** [comment]
Apologies @1ec5 - this was part of a PR creation batch and I should've paid more attention to this not being a recently created issue of mine.

**Comment by brunoabinader** [comment]
You are correct @jfirebaugh - we were actually doing the wrong coordinate system conversions in `SymbolAnnotationImpl::updateLayer` - so in fact #5419 is just a different manifestation of this issue :tada: :tada: :tada:

Going from `Point<double>` in `SymbolAnnotation` to `GeometryCoordinate` involves:
1. Convert `Point<double>` (aka. `LonLat`) to `TileCoordinate` via `TileCoordinate::fromLatLng`
2. Convert `Point<double>` from `TileCoordinate.p` to `GeometryCoordinate` (aka. `TilePoint`) via `TileCoordinate::toGeometryCoordinate`

For 2), we require #6626 - thus I'm rebasing this PR against that. We can either merge #6626 first and rebase this against that or just review and merge this and close #6626, I'm fine with either option.

**Comment by brunoabinader** [comment]
Fixes #5419.

**Comment by brunoabinader** [comment]
Because features can now appear in more than one tile when they intersect with other tile edges, `queryRenderedFeatures` would return wrong values. With https://github.com/mapbox/mapbox-gl-native/pull/3563/commits/a3c3f781e1219345afdea2774343c325a39f15e5 we now filter these results by removing features with the same identifier.

**Comment by brunoabinader** [comment]
Filtering out duplicated annotations affects the results of some query render tests e.g. `edge-cases/null-island` now outputs 1 feature instead of 4. I believe we shouldn't output duplicated features and thus, update the expectations for these.

**Comment by brunoabinader** [comment]
Another side-effect: opaque annotations on tile edges are overdrawn:
screen shot 2016-10-09 at 9 20 47 pm

**Comment by brunoabinader** [comment]
One solution to the opaque feature overdraw is to always enable symbol bucket clipping:
screen shot 2016-10-10 at 10 20 11 am

This potentially fixes #1673 and #5055.

**Comment by brunoabinader** [comment]
> This potentially fixes #1673 and #5055.

Not true for #1673. Features that are not exactly on the tile edges, but close enough are still clipped as you can see below (notice the cross pattern):
screen shot 2016-10-10 at 10 48 56 am

**Comment by brunoabinader** [comment]
i've had an insightful talk with @kkaefer today about the possibility to get rid of stencil clipping for symbols. In sum, these are the options we have so far:

# Option 1: Get rid of stencil clipping
- Filter symbol features in which geometries intersects more than one tile, so only one feature gets inside symbol layer
 - Implies disabling stencil tests
- __Pros__:
  - Fixes collision tile rendering
  - Fixes overdrawn features on tile edges
- __Cons__:
  - Because `symbol-avoid-edges` depends on the stencil clip, it stops working
  - 5 failing render-tests:
    - symbol-avoid-edges/default
    - symbol-avoid-edges/rotated-false
    - symbol-spacing/line-close
    - symbol-spacing/line-far
    - text-tile-edge-clipping/default

# Option 2: Always clip symbol buckets
- This is the default implementation for `MapMode::Still`
- __Pros__:
 - All tests pass
 - Fixes collision tile rendering
- __Cons__:
 - [Clipped features on tile edges](https://github.com/mapbox/mapbox-gl-native/pull/3563#issuecomment-252554312)
 - `queryRenderedFeatures` returns duplicated features - this is desired (?)
  - See e.g. https://github.com/mapbox/mapbox-gl-test-suite/blob/master/query-tests/edge-cases/null-island/expected.json
  - Depends on stencil clip - which @kkaefer suggested to get rid of in the future

**Comment by brunoabinader** [comment]
Pinging @jfirebaugh for evaluating the options.

**Comment by jfirebaugh** [comment]
Pretty sure that for now we want to go with Option 2: Always clip symbol buckets. Getting rid of stencil clipping for symbols is a good long-term goal but there are a number of tricky issues that we will need to solve for that to be viable, and we should fix this bug in the meantime.

**Comment by brunoabinader** [comment]
One way of fixing the clipped features close to tile edges is [increasing the bounds when querying for renderable features within tile bounds](https://github.com/mapbox/mapbox-gl-native/blob/master/src/mbgl/annotation/annotation_manager.cpp#L164-L167) - downside is that we'll eventually insert features to render tiles that won't be visible at all due to stencil clipping.

**Comment by brunoabinader** [comment]
Per conversion with @jfirebaugh, I've decided to split the fixes previously contained in this PR in separate ones:
- [core] Skip duplicated IDs when querying point annotations - #6655 
- [core] Cleanup SymbolLayout - #6666 
- [core] Disable stencil test when rendering collision boxes - #6669

The [commit left for review on this PR](https://github.com/mapbox/mapbox-gl-native/pull/3563/commits/dc81a13bc0f3f3b7173b8f5134e5994dbaa00dec) fixes the immediate cause for the issue initially mentioned on this issue (now PR).

In order to [fix rendering of opaque symbols near tile edges](https://github.com/mapbox/mapbox-gl-native/pull/3563#issuecomment-252503046), I'm moving this discussion to a new issue: #6670

**Comment by jfirebaugh** [comment]
This introduces a `TransformState` to a bunch of functions in order to satisfy a composition of functions -- essentially `TileCoordinate::toGeometryCoordinate(TileCoordinate::fromLatLng(latLng))` -- in which `TransformState` should "cancel out" and be irrelevant to the result.

Can we instead write a method such as `GeometryCoordinate CanonicalTileID::project(const LatLng&)` which does the geographic coordinate ⇢ tile coordinate projection directly? (It doesn't necessarily have to be a member of `CanonicalTileID` -- a standalone function would also be fine.)

**Comment by brunoabinader** [comment]
> Can we instead write a method such as GeometryCoordinate CanonicalTileID::project(const LatLng&) which does the geographic coordinate ⇢ tile coordinate projection directly?

Totally @jfirebaugh - this actually ended up in moving `{un,}project()` into `Projection` class - cleaning up a lot of code in between and moving us one step closer into switchable projection types in the future 😉

File: src/mbgl/util/tile_cover.cpp [hunk 1]
```diff
@@ -149,11 +149,11 @@ std::vector<UnwrappedTileID> tileCover(const LatLngBounds& bounds_, int32_t z) {
 
     const TransformState state;
```
Comment (jfirebaugh) [inline src/mbgl/util/tile_cover.cpp]: Unused.

**Comment by 1ec5** [comment]
This would be nice to get on the new release-ios-v3.4.0 branch before it diverges.
